serialization.rst

```diff
@@ -0,0 +1,337 @@
+    }
+  } => { 'foo' : 42, 'prot' : "wine", '__pclass' : Binary(0x80, "UpperClass") }
+
+
+Deserialization from BSON
+=========================
+
+For compound types, there are three data types:
+
+- ``root``: refers to the top-level BSON document *only*
+- ``document``: refers to nested BSON document *only*
+- ``array``: refers to a BSON array
+
+Each of those three data types can be mapped against different PHP types. The
+possible mapping values are:
+
+- *not set* or ``NULL`` — this is the default.
+
+  - A BSON array will be deserialized as a PHP ``array``.
+  - A BSON document (root or nested) without ``__pclass`` property [1]_ becomes a
+    PHP ``stdClass`` object, with each BSON document property becoming a
+    public ``stdClass`` property.
+  - A BSON document (root or nested) with ``__pclass`` property [1]_ becomes
+    a PHP object of the class name as defined by the ``__pclass`` property.
+
+    If the named class implements the ``MongoDB\BSON\Unserializable``
+    interface, then the properties of the BSON document, minus the
+    ``__pclass`` property are sent as an associative array to the
+    ``bsonUnserialize`` function to initialise the object's properties.
+    
+    If the named class does not implement the ``MongoDB\BSON\Unserializable``
+    interface, then the object gets initialized with the properties
+    **without** the ``__pclass`` property being one of them. It simply gets
+    left out.
+
+    The rationale for the latter is because no explicit choice about whether
+    to use ODM features has been made, it should not bail out because of an
+    extra ``__pclass`` property. The behaviour is different if the data type
+    was mapped to ``"odm"`` (see below).
+
+- ``"array"`` — turns a BSON array or BSON document into a PHP array.
+  ``__pclass`` properties [1]_ are igored.
+
+- ``"object"`` or ``"stdClass"`` — turns a BSON array or BSON document into a
+  ``stdClass`` object. There will be no special treatment of a ``__pclass``
+  property [1]_, buy it should **not** be set as property in the returned
+  object.
+
+- ``"odm"`` — uses the class name set in the ``__pclass`` property of a BSON
+  array or BSON document to determine the class name of the created object.
+
+  The properties of the BSON document, minus the ``__pclass`` property are
+  sent as an associative array to the ``bsonUnserialize`` function to initialise
+  the object's properties.
+
+  Remarks:
+
+  - If the ``__pclass`` property is not present [1]_, this throws the
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating the special class marker could not be found.
+
+  - If the class as determined by the ``__pclass`` property does not implement
+    the ``MongoDB\BSON\Persistable`` interface, throw an
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating that the named class does not implement the required
+    interface.
```

Comment:
Yes, I believe this was addressed in [PHPC-274](https://jira.mongodb.org/browse/PHPC-274). Looking at the PR in 10gen-labs/mongo-php-driver-prototype#40, I changed this so that we simply check for `Serializable` to get our data to serialize via `bsonSerialize()`. `Persistable` only comes into play to inject the `__pclass` value (if the `PHONGO_BSON_ADD_ODS` flag is also set).

What is a slippery slope here? I think the main points are:

 * During encoding: if a class is `Serializable`, we use `Serializable::bsonSerialize()` to get an associative array of fields and values for our BSON document.
   * If the `Serializable` class is also `Persistable`, and our state flag allows it, we'll inject a `__pclass` field into the document (possibly overwriting a `__pclass` field that may have existed in the array returned by `Serializable::bsonSerialize()`.
 * During decoding: if a class is `Unserializable`, we pass an associative array of fields and values from the BSON document to `Serializable::bsonSerialize()` (I'm omitting discussion of how we find the class based on the type map and `__pclass` here).